navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/instruction/DefaultInstructionEngine.java

```diff
@@ -0,0 +1,223 @@
+      return buildThenFormatInstruction(progress);
+    } else {
+      return progress.getCurrentLegProgress().getUpComingStep().getManeuver().getInstruction();
+    }
+  }
+
+  /**
+   * On arrival, use the upcoming step instruction.
+   * If empty, use the current step instruction as a fallback
+   * @param progress {@link RouteProgress} created by the location change
+   * @return {@link String} to be announced on departure milestone
+   */
+  private String buildArrivalInstruction(RouteProgress progress) {
+    if (progress.getCurrentLegProgress().getUpComingStep() != null) {
+      String instruction = progress.getCurrentLegProgress().getUpComingStep().getManeuver().getInstruction();
+      if (!TextUtils.isEmpty(instruction)) {
+        return instruction;
+      }
+    } else {
+      return progress.getCurrentLegProgress().getCurrentStep().getManeuver().getInstruction();
+    }
+    return "";
+  }
+
+  /**
+   * Creates a {@link String} with the current step name and distance remaining
+   * Example: "Continue on Main St. for 3.2 miles"
+   * @param progress {@link RouteProgress} created by the location change
+   * @return {@link String} with format "Continue on %s for %s"
+   */
+  private String buildContinueFormatInstruction(RouteProgress progress) {
+    double userDistance = progress.getCurrentLegProgress().getCurrentStepProgress().getDistanceRemaining();
+    return String.format(
+      Locale.US,
+      CONTINUE_STRING_FORMAT,
+      progress.getCurrentLegProgress().getCurrentStep().getName(),
+      distanceFormatter(userDistance));
+  }
+
+  /**
+   * Creates a {@link String} with the current step distance remaining upcoming step instruction
+   * Example: "In 3.2 miles turn left onto Main St."
+   * @param progress {@link RouteProgress} created by the location change
+   * @return {@link String} with format "In %s %s"
+   */
+  private String buildDefaultFormatInstruction(RouteProgress progress) {
+    double userDistance = progress.getCurrentLegProgress().getCurrentStepProgress().getDistanceRemaining();
+    return String.format(
+      Locale.US,
+      IN_STRING_FORMAT,
+      distanceFormatter(userDistance),
+      convertFirstCharLowercase(progress.getCurrentLegProgress()
+        .getUpComingStep().getManeuver().getInstruction())
+    );
+  }
+
+  /**
+   * Creates a {@link String} with the current step maneuver instruction, current step distance remaining,
+   * and upcoming step instruction
+   * Example: "Turn left onto Main St. then in 3.2 miles turn right onto Second St."
+   * @param progress {@link RouteProgress} created by the location change
+   * @return {@link String} with format "%s then in %s %s"
+   */
+  private String buildThenInFormatInstruction(RouteProgress progress) {
+    double userDistance = progress.getCurrentLegProgress().getCurrentStepProgress().getDistanceRemaining();
+    StepManeuver currentStepManeuver = progress.getCurrentLegProgress().getCurrentStep().getManeuver();
```

Comment:
might be cleaner to change to `String currentStepInstruction = progress.getCurrentLegProgress().getCurrentStep().getManeuver().getInstruction();`